crates/red_knot_python_semantic/src/types/class.rs

```diff
@@ -28,6 +32,160 @@ use super::{
     TypeQualifiers, TypeVarInstance,
 };
 
+/// Tries to find declarations/bindings of an instance attribute named `name` that are only
+/// "implicitly" defined in a method of the class that corresponds to `class_body_scope`.
+#[allow(clippy::needless_pass_by_value)]
+#[salsa::tracked]
```

Comment:
Have you investigated where the performance regression is coming from? 

I suspect that it is because this function is now a salsa query and its arguments are somewhat expensive (requires hashing and allocating a name). Gating the `attribute_assignments` and `is_attribute_assignment_visible` methods behind a query is important because they access `semantic_index` or `node`. However, it makes me wonder if we could change the methods so that they wouldn't have to depend on the entire `semantic_index` or access specific nodes.

Reply:
Thanks. Initially I thought the performance regression was caused by removing `salsa::tracked` of `attribute_assignments`, but changing the field types of `{Assignment, ForStmt, WithItem}DefinitionKind` may be the true cause (there seems to be a regression with type inference on these).